py_mini_racer/extension/mini_racer_extension.cc

```diff
@@ -177,21 +186,28 @@ static void gc_callback(Isolate *isolate, GCType type, GCCallbackFlags flags) {
 }
 
 static void init_v8() {
+    // no need to wait for the lock if already initialized
+    if (current_platform != NULL) return;
+
+    platform_lock.lock();
+
     if (current_platform == NULL) {
         V8::InitializeICU();
-        current_platform = platform::CreateDefaultPlatform();
-        V8::InitializePlatform(current_platform);
+        current_platform = platform::NewDefaultPlatform();
+        V8::InitializePlatform(current_platform.get());
         V8::Initialize();
     }
+
+    platform_lock.unlock();
 }
 
-static void* breaker(void *d) {
+static void breaker(std::timed_mutex& breaker_mutex, void * d) {
   EvalParams* data = (EvalParams*)d;
-  usleep(data->timeout*1000);
-  pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL);
-  data->result->timed_out = true;
-  data->context_info->isolate->TerminateExecution();
-  return NULL;
+
+  if (!breaker_mutex.try_lock_for(std::chrono::milliseconds(data->timeout))) {
```

Comment:
Is this cancellable?

Reply:
Yep, the mutex is relaxed when the execution is done, otherwise it will timeout and exit.